### PR TITLE
Add miniver to simplify development branch descriptions

### PR DESCRIPTION
**Pull request by hmaarrfk**
Closes: https://github.com/pygfx/pygfx/issues/696

```bash
pip install -e . -vv
# You might only need to run this once I believe. we don't run it daily on our dev machines
```

```python
In [1]: import pygfx

In [2]: pygfx.__version__
'0.1.18.post29+g9342016'

In [3]: from packaging.version import Version

In [4]: Version('0.1.18.post29+g9342016') > Version('0.1.18')
True

In [5]: assert Version('0.1.18.post29+g9342016') > Version('0.1.18')

In [6]: assert Version('0.1.18.post29+g9342016') < Version('0.1.19')
```

So I use this in my project and the github action that I use to upload the package to pypi will make the static version correctly.

You don't need to change the source when you create a new release. just tag then your CI will pickup the new version in the upload to pypi.

see: https://github.com/ramonaoptics/python-teensytoany
Xref: https://github.com/jbweston/miniver

### CODE REVIEW

File: pygfx/__init__.py [hunk 1]
```diff
@@ -25,8 +25,8 @@
 
 del pylinalg
 
-__version__ = "0.1.18"
-version_info = tuple(map(int, __version__.split(".")))
+from ._version import __version__
+version_info = tuple(map(int, __version__.split(".")[:3]))
```
Comment (hmaarrfk) [inline pygfx/__init__.py:29]: I'm not sure what behavior you want from this `version_info` tuple. I don't have such constructs in my code.

I typically use

```
from packaging import Version
```
for all my comparisons.

**Comment by hmaarrfk** [comment]
oh man, the CIs made this more difficult than I wanted to.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -41,7 +44,10 @@ jobs:
     strategy:
       fail-fast: false
     steps:
-    - uses: actions/checkout@v3
+    - uses: actions/checkout@v4
```
Comment (hmaarrfk) [inline .github/workflows/ci.yml:47]:
```suggestion
    - uses: actions/checkout@v3
```

**Comment by hmaarrfk** [comment]
hmm, i've never used pyinstaller ...

**Comment by hmaarrfk** [comment]
This is actually the first time i read through the miniver code in earnest.

I understand how it is designed now, and I somewhat understand their design decisions of having 2 files.

But it is somewhat overly complicated.

It may be interesting to use https://github.com/proudust/gh-describe though this seems like getting more untrusted code in the repo.

File: pygfx/__init__.py (hunk 1 above)
Comment (almarklein) [reply pygfx/__init__.py:30]: Let's do this:
```suggestion
version_info = tuple(int(v) if v.isnumeric() else v for v in __version__.split("."))
```

File: pygfx/_version.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+def get_version_from_git():
+    import subprocess
+
+    # git describe --first-parent does not take into account tags from branches
+    # that were merged-in. The '--long' flag gets us the 'post' version and
+    # git hash, '--always' returns the git hash even if there are no tags.
+    for opts in [["--first-parent"], []]:
+        try:
+            p = subprocess.Popen(
+                ["git", "describe", "--long", "--always", "--tags", "--dirty"] + opts,
```
Comment (almarklein) [inline pygfx/_version.py:50]: IIUC this means that if someone obtains pygfx in another way than via a package manager or git repo, the version cannot be established.

**Review by almarklein — COMMENTED** [review]
I'm not overly enthusiastic 😅 but this may also be because I don't understand the problem that it solves. Let's discuss in #696

**Comment by Korijn** [comment]
My general feeling about this is that it's adding a ton of code and complexity to solve a very small simple manual step... not too enthusiastic.

File: pygfx/_version.py (hunk 3 above)
Comment (hmaarrfk) [reply pygfx/_version.py:50]: Correct. There become some "edge cases" that you have to start taking into account when people "receive" the source code. 

I just checked my project for example, the version downloaded from pypi burns things in due to the modifications to the setup.py but the archive from github would make it pretty difficult for people to use.

image

**Comment by hmaarrfk** [comment]
> complexity and dependencies [...]  not too enthusiastic.

Understood. This is a fair criticism.

**Pull request closed without merging**